app/javascript/components/widget-display.jsx

```diff
@@ -0,0 +1,207 @@
+import React, { useRef, useEffect, useReducer } from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { mathMod } from 'ramda';
+import { DisconnectedMessage, LoadingMessage } from '@messages/message';
+import FullPanel from '@components/full-panel';
+import SidePanel from '@components/side-panel';
+import Twitter from '@widgets/twitter';
+import Numbers from '@widgets/numbers';
+import Weather from '@widgets/weather';
+import Traffic from '@widgets/traffic';
+
+const Wrapper = styled.div`
+  outline: none;
+`;
+
+const widgetElements = {
+  Weather,
+  Twitter,
+  Numbers,
+  Traffic,
+};
+
+const prefetchDelay = 1500;
+let lastPrefetchTime = 0;
+
+const reducer = (state, action) => {
+  const { timerStart, timerValue } = state;
+  const now = Date.now();
+
+  switch (action.type) {
+    case 'togglePause':
+      return {
+        timerValue: timerStart ? timerValue - (now - timerStart) : timerValue,
+        timerStart: timerStart ? null : now,
+      };
+    case 'startTimer':
+      return {
+        timerStart: timerStart ? now : null,
+        timerValue: action.durationSeconds,
+      };
+    default:
+      return state;
+  }
+};
+
+export const WidgetDisplay = ({
+  requestPrefetch,
+  requestWidget,
+  cityName,
+  widgets,
+}) => {
+  const { byIdOrFirst: current, enabled: enabledWidgets } = widgets;
+
+  const widgetTransitionTime = current.durationSeconds * 1000;
+
+  const WidgetElement = widgetElements[current.name].Panel;
+
+  const index = enabledWidgets.findIndex(w => w.id === current.id);
+
+  const prevWidgetId =
+    enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
+  const nextWidgetId =
+    enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
+
+  const [state, dispatch] = useReducer(reducer, {
+    timerStart: widgetTransitionTime,
+    timerValue: null,
+  });
+
+  const timerId = useRef(null);
+
+  useEffect(() => {
+    dispatch({
+      type: 'startTimer',
+      durationSeconds: widgetTransitionTime,
+    });
+  }, [current.id]);
+
+  useEffect(() => {
+    timerId.current = setInterval(() => {
```

Comment:
```js
const timerId = ...
```
Since we're not using this ref outside of the effect we don't need to create a ref at all. We can store it in a const and clean it up in the return-callback.